### ISSUE TITLE
How to train properly from scratch?

### ISSUE THREAD

**Issue by jozefmorvay** [pierluigiferrari/ssd_keras#146]
The instructions in training notebook are crystal clear, however, you only seem to train on "07+12" trainval, and evaluate on 07 test. 

How do I reproduce the results from the main README.md that even surpass the original? They claim to have first trained on COCO and then finetuned on VOC 07+12, but I can't figure out if the model was trained on COCO to convergence, or simply for a certain number of epochs and finetuning was performed on "interim" weights. How exactly did you do it? I would like to know exact number of epochs for both datasets.

The original:
![image](https://user-images.githubusercontent.com/4798190/42049666-e67f4d04-7b06-11e8-8ed9-02219f542dfe.png)

Your implementation:
![image](https://user-images.githubusercontent.com/4798190/42049770-4806b6ac-7b07-11e8-8e77-32523843100a.png)

 I plan on doing this to test a feature-reinjection to help SSD better recognize smaller objects from [https://arxiv.org/pdf/1712.00960.pdf](url)

I need the model to recognize "person" class once trained, so I believe training on both COCO and VOC can only help.

**Comment by pierluigiferrari** [comment]
> The instructions in training notebook are crystal clear, however, you only seem to train on "07+12" trainval, and evaluate on 07 test.

Correct, that is the only training I replicated. The point of replicating the training of one model configuration was just to test whether my implementation is accurate. There was no need to replicate the trainings of the other configurations, because I already provide all the ported weights for all the original model configurations.

The results you're quoting are from the ported weights, not from my own training.

If you wanted to reproduce the trainings of those models that were trained on both MS COCO and Pascal VOC, you'd have to consult the original Caffe implementation for instructions. All the models there come with Python training scripts. Although presumably they just trained the respective model to convergence on MS COCO, then sub-sampled the classifier weights in the canonical way (the Pascal VOC classes are a subset of the MS COCO classes), and then trained to convergence on Pascal VOC.

> I need the model to recognize "person" class once trained, so I believe training on both COCO and VOC can only help.

Yes, but it makes an important difference whether you train on two datasets sequentially or whether you train on the union of both simultaneously. MS COCO is a much larger dataset with more objects per image on average, more different classes, and more smaller objects. In that sense it is a more complex and thus more challenging dataset. Training on this more complex dataset first and then training on a similar but less complex dataset (Pascal VOC) afterwards might not necessarily lead to a general improvement, but rather to a deterioration, depending on your specific application. The network might "unlearn" some of the structure that was necessary for MS COCO, but is no longer necessary during the training on Pascal VOC. I would assume that merging both datasets and training on the union would generally be a better idea. Those are both just my assumptions though, I haven't tried this and might be wrong.

**Comment by jozefmorvay** [comment]
So the original learning rate and number of epochs is mentioned in the original SSD paper. The actual python training script I was able to find at [https://github.com/weiliu89/caffe/blob/master/python/train.py](url) is not very descriptive, as it parses command line arguments and no values appear to be hard-set. 

I have two more points I'd like to have clarified, if you would be so kind. 

1. Do you have any insight into why a single epoch is 1000 steps? With 32 images in batch, that gives you 32000 images per epoch, but the 07+12 train set only has 16551 images. It appears almost as if this value was chosen arbitrarily. Is there an actual meaning to it?

2. in your training notebook, you load VGG detection weights but don't freeze them. The additional SSD untrained randomly-initialized layers will probably have large updates by the optimizer. Will not such updates negatively affect the already trained layers in the VGG, since they propagate all the way to the first conv layers?

**Comment by pierluigiferrari** [comment]
> So the original learning rate and number of epochs is mentioned in the original SSD paper. The actual python training script I was able to find at https://github.com/weiliu89/caffe/blob/master/python/train.py is not very descriptive, as it parses command line arguments and no values appear to be hard-set.

The training scripts are not on GitHub. You have to download the models. The training scripts are in the zip files for the respective models.

> 1. Do you have any insight into why a single epoch is 1000 steps? With 32 images in batch, that gives you 32000 images per epoch, but the 07+12 train set only has 16551 images. It appears almost as if this value was chosen arbitrarily. Is there an actual meaning to it?

The concept of an epoch is utterly meaningless with regard to the training performance. It never matters how many steps one epoch consists of, it only matters for how many steps the model has been trained in total and how many times it has seen the dataset during the training (i.e. the number of passes over the dataset). One epoch is 1000 steps in the SSD300 training notebook because I chose it to be 1000 steps. It would make no difference if it were 100 steps or 469 steps. The relevant number is that the "07+12" model was trained for (around) 120,000 steps. How these steps are split up across epochs is unimportant for the training itself.

> 2. in your training notebook, you load VGG detection weights but don't freeze them.

Correct. And neither does the original Caffe implementation. I wouldn't freeze any weights unless I was concerned about overfitting (e.g. if I had a very small dataset).

> The additional SSD untrained randomly-initialized layers will probably have large updates by the optimizer. Will not such updates negatively affect the already trained layers in the VGG, since they propagate all the way to the first conv layers?

This isn't really the right place for this question, but the derivatives at a given layer depend on the inner product of the derivatives at the subsequent layer and the weights between the two layers, so how much of the gradient gets propagated depends on this inner product.

**Comment by jozefmorvay** [comment]
If I want to merge COCO and VOC datasets, do I still have to consider different scales for each dataset? Scales are only used to scale box_height and box_width, and do not partake in loss function, where only offsets are considered (line taken from SSDLoss class):

localization_loss = tf.to_float(self.smooth_L1_loss(y_true[:,:,-12:-8], y_pred[:,:,-12:-8]))

Those indexes strip away class predictions, variances and anchor box coordinates, if I understand correctly, and only leave offsets.

Is my understanding correct, that during training, the scales are irrelevant?

**Comment by pierluigiferrari** [comment]
> Is my understanding correct, that during training, the scales are irrelevant?

No, that is not correct. If the scales weren't relevant for the training, then what else would they be relevant for?

The scaling factors determine the sizes of the anchor boxes. The sizes of the anchor boxes affect which ground truth boxes get matched to which anchor boxes. This matching process determines what the individual neurons in the predictor layers will learn. That is, the scaling factors are very relevant to the training.

**Comment by jozefmorvay** [comment]
So how can I train on both datasets at the same time?

**Comment by pierluigiferrari** [comment]
Same as you would treat any dataset: You pick scaling factors that cover the object sizes in your dataset well.

In the case of a merged MS COCO & Pascal VOC dataset that means just taking the MS COCO scaling factors (which you could also have figured out from the fact that the Pascal VOC "07+12+COCO" models use the MS COCO scaling factors and work very well).

**Comment by jozefmorvay** [comment]
Thank you, yet again. One additional question, what use are variances? The paper never mentions them, but they are in the code and I don't quite understand why you should scale the anchors like that.

**Comment by pierluigiferrari** [comment]
You can interpret them as the assumed variances of the distributions of the offsets from the respective anchor box coordinates (not to be confused with the anchor box parameter named `offsets`; here I just mean offsets as in "deviations" from the anchor box coordinates). From this perspective, dividing by them normalizes those distributions. Or, if you don't know much about probability theory, you can just see them as scaling factors for the anchor box offsets. They allow you to scale the offsets and, in particular, to scale the individual coordinate offsets differently.

I have not studied their effect and, as you mentioned, the authors of the paper provide no justification as to why they introduced these variances, let alone why they chose their values the way they did.

**Comment by cyberphantom** [comment]
I think by offsets you mean aspect rations in your implementation. Can you please clarify why the offset values are 0.5 for all the layers? I read the comment but I can't imagine how it works
